pkg/querier/batch/stream.go

```diff
@@ -42,61 +50,95 @@ func (bs *batchStream) at() (int64, float64) {
-			// Add another batch the the result
-			// and use it for further appending.
-
-			// The Index is the place at which new sample
-			// has to be appended, hence it tells the length.
-			b.Length = b.Index
-			resultLen++
-			if resultLen > len(result) {
-				// It is possible that result can grow longer
-				// then the one provided.
-				result = append(result, chunk.Batch{})
-			}
-			b = &result[resultLen-1]
+	// Step to the next Batch in the result, create it if it does not exist
+	nextBatch := func(valueType chunkenc.ValueType) {
+		// The Index is the place at which new sample
+		// has to be appended, hence it tells the length.
+		b.Length = b.Index
+		resultLen++
+		if resultLen > len(result) {
+			// It is possible that result can grow longer
+			// then the one provided.
+			result = append(result, chunk.Batch{})
+		}
+		b = &result[resultLen-1]
+		b.ValueType = valueType
+	}
+
+	populate := func(s batchStream, valueType chunkenc.ValueType) {
+		if b.Index == 0 {
+			// Starting to write this Batch, it is safe to set the value type
+			b.ValueType = valueType
+		} else if b.Index == size || b.ValueType != valueType {
+			// The batch reached its intended size or is of a different value type
+			// Add another batch to the result and use it for further appending.
+			nextBatch(valueType)
 		}
+
+		switch valueType {
+		case chunkenc.ValFloat:
+			b.Timestamps[b.Index], b.Values[b.Index] = s.at()
+		case chunkenc.ValHistogram:
+			t, v := s.atHistogram()
+			b.Timestamps[b.Index], b.PointerValues[b.Index] = t, unsafe.Pointer(v)
+		case chunkenc.ValFloatHistogram:
+			t, v := s.atFloatHistogram()
+			b.Timestamps[b.Index], b.PointerValues[b.Index] = t, unsafe.Pointer(v)
+		}
+		b.Index++
 	}
 
-	for left.hasNext() && right.hasNext() {
-		checkForFullBatch()
+	for lt, rt := left.hasNext(), right.hasNext(); lt != chunkenc.ValNone && rt != chunkenc.ValNone; lt, rt = left.hasNext(), right.hasNext() {
 		t1, t2 := left.atTime(), right.atTime()
 		if t1 < t2 {
-			b.Timestamps[b.Index], b.Values[b.Index] = left.at()
+			populate(left, lt)
 			left.next()
 		} else if t1 > t2 {
-			b.Timestamps[b.Index], b.Values[b.Index] = right.at()
+			populate(right, rt)
 			right.next()
 		} else {
-			b.Timestamps[b.Index], b.Values[b.Index] = left.at()
+			if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
+				// Prefer histograms over floats. Take left side if both have histograms.
+				populate(right, rt)
+			} else {
+				populate(left, lt)
+			}
 			left.next()
 			right.next()
 		}
-		b.Index++
 	}
 
 	// This function adds all the samples from the provided
 	// batchStream into the result in the same order.
 	addToResult := func(bs batchStream) {
-		for ; bs.hasNext(); bs.next() {
-			checkForFullBatch()
-			b.Timestamps[b.Index], b.Values[b.Index] = bs.at()
-			b.Index++
+		for t := bs.hasNext(); t != chunkenc.ValNone; t = bs.hasNext() {
+			populate(bs, t)
 			b.Length++
```

Comment:
Why do we need to increase `b.Length` here? It should be unnecessary because (a) what we want to increase is `b.Index` but it's already done by `populate()` and (b) `b.Length` is overridden later by

```
// The Index is the place at which new sample
// has to be appended, hence it tells the length.
b.Length = b.Index
```

Reply:
all checks pass if I remove it. not sure why it was there in the first place